MEN-4136: Update download package with options to use apt repository

alfrunes opened the pull request:
Adds two more installation paths:
 - Using the convenience script (https://get.mender.io)
 - Setting up our APT repository and installing from there.

(requires mendersoftware/mender-docs-site#297 to render properly)

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+
+* Express installation using the [convenience
+  script](#express-installation) from [https://get.mender.io](https://get.mender.io).
+* Set up Mender's apt repository and using the [package
```
I think we should capitalize `APT`

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+#### Express installation
+
+Mender provides a convenience script available at [get.mender.io
+](https://get.mender.io) that non-interactively install the Mender client
```
installs

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+using this option should be aware of:
+
+* The script require `root` privileges to run. Therefore, be careful to
+  examine the script before executing them.
```
executing it.

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+  examine the script before executing them.
+* The script install several dependencies with the package manager without
+  asking for confirmation.
+* The Mender GPG public key and apt repository will be silently added to your
```
APT

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+  examine the script before executing them.
+* The script install several dependencies with the package manager without
+  asking for confirmation.
+* The Mender GPG public key and apt repository will be silently added to your
```
And I wouldn't say "silently". I mean, is a script with standard verbosity. We could make it more verbose if necessary.

alfrunes replied:
I changed "silently" to "... without asking for confirmation".

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+
+* The script require `root` privileges to run. Therefore, be careful to
+  examine the script before executing them.
+* The script install several dependencies with the package manager without
```
The script will install

lluiscampos replied:
Maybe: "The script will install necessary Mender dependencies..."

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+* The script install several dependencies with the package manager without
+  asking for confirmation.
+* The Mender GPG public key and apt repository will be silently added to your
+  trusted apt keychain and sources list respectively.
```
APT

lluiscampos replied:
(more below, but I'll stop commenting on these now)

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+* The Mender GPG public key and apt repository will be silently added to your
+  trusted apt keychain and sources list respectively.
+
+If any of these issues are of concern to you, consider using one of the other
+installation methods.
```
In my personal opinion, this is redundant: we already have given other options and we have already warn extensively. I would just remove it

lluiscampos commented on 09.Downloads/docs.md:
````diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+!! Always examine scripts downloaded over the Internet before running them
+!! locally.
+```bash
+curl -fLsS https://get.mender.io -o get-mender.sh
+# INSPECT get-mender.sh BEFORE PROCEEDING
+sudo sh get-mender.sh
````
Okey, so no `curl -fLsS https://get.mender.io -o get-mender.sh | sudo bash` ? ;)

I envisioned a one-liner, but I think you are right in being as explicit as possible about the security risks.

alfrunes replied:
Yeah, I think it's better to just be honest and explicit about good security practices.

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+##### Upgrading Mender after the express installation
+
+When installing the Mender client with [get.mender.io](https://get.mender.io),
+the `mender-client` package is maintained by the package manager. *Do not* run
+this script multiple times to upgrade the Mender client. This will cause
+multiple additions of the Mender Debian repository to the `apt` sources list,
+which can cause issues when upgrading packages using apt.
```
I think you are right but I consider this a bug or weakness in the script. We can follow-up separately to make the script idempotent, do you agree?

alfrunes replied:
I think this is a trade-off. If we make the script more involved it's also making it harder for users to inspect it. I don't see it as a critical issue that users can't run this twice. In any case, I think we should encourage users to use `apt upgrade` from this point on somehow.

mchalski replied:
I'll make it idempotent, as discussed in https://github.com/mendersoftware/mender-dist-packages/pull/46

lluiscampos commented on 09.Downloads/docs.md:
````diff
@@ -83,8 +87,128 @@ software updates packaged as Mender Artifacts.
+           main"
+   ```
+   [/ui-tab]
+   [/ui-tabs]
````
This looks nice, but we are missing a note being explicit about choosing the correct architecture. Maybe "Add the Mender repository to your sources list. Chose the architecture matching your device."

An alternative could have been having it in a single script using `dpkg --print-architecture` instead of three different tabs.

lluiscampos reviewed (changes requested):
Thanks @alfrunes, looks quite good. A bunch of comments below.

alfrunes commented:
@lluiscampos I've addressed most of the comment and added a paragraph explaining the `mender-shell` dependency in the Express installation. I'll also add the rest of the "Remote terminal client" sub-section tomorrow. I'll hold off the next review request until this is done.

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,143 @@ software updates packaged as Mender Artifacts.
+When installing the Mender client with [get.mender.io](https://get.mender.io),
+the `mender-client` package is maintained by the package manager. *Do not* run
+this script multiple times to upgrade the Mender client. This will cause
+multiple additions of the Mender Debian repository to the APT sources list,
```
I would omit Debian, to avoid confusion with upstream repo

```suggestion
multiple additions of the Mender repository to the APT sources list,
```

alfrunes replied:
I'm ok with that if it does not lead to confusion with "GitHub repository".

lluiscampos replied:
Good point. "Mender APT repository", maybe? Does someone else have an opinion?

lluiscampos commented on 09.Downloads/docs.md:
```diff
@@ -101,6 +240,31 @@ Ubuntu or Raspberry Pi OS. We provide packages for the following architectures:
+the [express installation](#express-installation) script, you will already have
+`mender-shell` installed by default.
+
+The remote terminal client is only available from the Mender Debian repository.
```
Same suggestion here

```suggestion
The remote terminal client is only available from the Mender repository.
```

lluiscampos reviewed (approved):
Looks good to me!

The broken link I guess it because we need #1241 

Also, let's discuss the requirement or not of a feature branch in Slack `mender-docs-team` before merging this.

drewmoseley commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+* The Mender GPG public key and APT repository will be added to your trusted APT
+  keychain and sources list respectively without asking for confirmation.
+
+!! Always examine scripts downloaded over the Internet before running it
```
before running them locally.

drewmoseley reviewed (changes requested):
Looks good except for one minor change.  Definitely don't need to re-review for that fix alone.

merlin-northern commented on 09.Downloads/docs.md:
````diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+   		gnupg-agent \
+   		software-properties-common
+   ```
+2. Add the official Mender GPG to your trusted `apt` keychain:
````
suggested change "official Mender GPG to" -> "official Mender GPG _key_ to"

merlin-northern commented on 09.Downloads/docs.md:
````diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+specifying which packages you want to install. For example, the following will
+only install the Mender client:
+```bash
+curl -fLsS https://get.mender.io -o get-mender.sh
````
```
#curl -fLsS https://get.mender.io -o get-mender.sh
curl: (6) Could not resolve host: get.mender.io
```

alfrunes replied:
This URL depends on [MEN-4150](https://tracker.mender.io/browse/MEN-4150) and [MEN-4152](https://tracker.mender.io/browse/MEN-4152) - hence, the feature branch.

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+
+##### Upgrading Mender after the express installation
+
+After installing the Mender client with [get.mender.io](https://get.mender.io),
```
suggested change: 
either
"installing the Mender client with [get.mender.io]..." -> "installing the Mender client _from_ [get.mender.io](https://get.mender.io)"
or
"installing the Mender client with [get.mender.io]..." -> "installing the Mender client _with_ get-mender.sh "

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+##### Upgrading Mender after the express installation
+
+After installing the Mender client with [get.mender.io](https://get.mender.io),
+the `mender-client` package is managed by the package manager. To upgrade the
```
managed by a manager, and also passive. maybe somehow reword this? "you can upgrade mender-client in a standard way with the package manager:"

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+
+#### Install using the APT repository
+
+Before installing the Mender client for the first time on a new system, you need
+to set up the Mender APT repository. Afterwards, you can install and update
+the Mender client from the repository.
```
Second sentence is redundant. I would suggest some rewording of this paragraph:
"If you choose to install the Mender client with the `apt-get` command you need to first configure the APT repository."
just a suggestion, I somehow do not like how this reads.

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+
+#### Install from package
+
+The Debian package (`.deb`) is also provided for users wanting to install the
```
passive. suggested change: "We also provide the Debian package .."

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+  script](#express-installation) from [https://get.mender.io](https://get.mender.io).
+* Set up Mender's APT repository and using the [package
+  manager](#install-using-the-apt-repository).
+* Download our DEB package and [install it manually](#install-from-package).
```
this is the one and only place where we use "`DEB`" string, in the rest of this section we refer to the package as "Debian package" with occasional ".deb".

merlin-northern replied:
Technically speaking, you are listing here three "different ways" to install. First one "Express installation" is a way to install. second and third "Set up Mender's APT" and "Download our DEB" could be rephrased to better fit here. e.g.: "Download our DEB" -> "Install from a .deb file", "Set up Mender's APT" -> "Install from the APT repository"

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+[using the package manager](#install-using-the-apt-repository). Users considering
+using this method should be aware of:
+
+* The script require `root` privileges to run. Therefore, be careful to
```
suggested change: "The script require " -> "The script _requires_ "

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+Mender provides a convenience script available at [get.mender.io
+](https://get.mender.io) that non-interactively installs the Mender client
+[using the package manager](#install-using-the-apt-repository). Users considering
+using this method should be aware of:
```
suggested change "be aware of:" -> "be aware of _the following_:" or something in the lines of "aware that:"

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+using this method should be aware of:
+
+* The script require `root` privileges to run. Therefore, be careful to
+  examine the script before executing it.
```
suggested change: "examine the script before executing it." -> "examine it before executing."

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+
+The Debian package (`.deb`) is also provided for users wanting to install the
+Mender client manually outside the package manager. This can be useful for
+airtight systems with limited access to the Internet, for example when running
```
somehow this does not read well.
suggested change: "access to the Internet, for example when running" -> "access to the Internet, _or_ when running"

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -83,8 +87,145 @@ software updates packaged as Mender Artifacts.
+The Debian package (`.deb`) is also provided for users wanting to install the
+Mender client manually outside the package manager. This can be useful for
+airtight systems with limited access to the Internet, for example when running
+Mender in [standalone
```
suggested change: "Mender in [standalone" -> "Mender in _the_ [standalone"

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -101,6 +242,31 @@ Ubuntu or Raspberry Pi OS. We provide packages for the following architectures:
+accessing the device terminal using the Mender API. See the [configuration page
+for remote
+terminal](../03.Client-installation/80-Mender-shell-configuration/docs.md) for
+more information about the remote terminal plugin.
```
suggested change: drop: " about the remote terminal plugin"

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -101,6 +242,31 @@ Ubuntu or Raspberry Pi OS. We provide packages for the following architectures:
 [mender-client_x.x.x-1_amd64.deb]: https://downloads.mender.io/master/dist-packages/debian/amd64/mender-client_master-1_amd64.deb
 
 
+## Remote terminal client
```
suggested: let's not use "remote terminal client" every where we use mender-shell or add-on, so maybe change to "Remote terminal add-on"?

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -101,6 +242,31 @@ Ubuntu or Raspberry Pi OS. We provide packages for the following architectures:
+## Remote terminal client
+
+Mender offers a remote terminal extension to the Mender client that enables
+accessing the device terminal using the Mender API. See the [configuration page
```
we do not provide any API for remote terminal.
suggested change: "using the Mender API" -> "using the Mender _UI_"

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -101,6 +242,31 @@ Ubuntu or Raspberry Pi OS. We provide packages for the following architectures:
+
+### Install the remote terminal client
+
+The remote terminal client requires installing [Mender client](#mender-client)
+in order to function, so make sure you have installed the Mender client before
+installing `mender-shell`. If you have already installed the Mender client using
+the [express installation](#express-installation) script, you will already have
+`mender-shell` installed by default.
```
this does not read very well.
suggested change:
"The mender-shell add-on requires the Mender client in order to provide remote terminals to accepted devices from the Mender UI. Please also note, that if you used [the express installation method](#express-installation), you will already have
`mender-shell` installed by default."

merlin-northern commented on 09.Downloads/docs.md:
```diff
@@ -101,6 +242,31 @@ Ubuntu or Raspberry Pi OS. We provide packages for the following architectures:
+the [express installation](#express-installation) script, you will already have
+`mender-shell` installed by default.
+
+The remote terminal client is only available from the Mender APT repository.
+To install `mender-shell`, follow the instructions for [installing
+`mender-client` using the APT repository](#install-using-the-apt-repository).
+After the final step, install `mender-shell` using the package manager:
```
suggested change:
"The mender-shell add-on is only available from the Mender APT repository. Once you perform [the steps to configure the APT repository](#install-using-the-apt-repository), you can go with the installation in a normal way, by running:"

merlin-northern reviewed (commented):
left some comments.

merlin-northern reviewed (approved):
I trust you -- did not revisit all.

alfrunes commented:
I'm merging this into the feature branch. We may revisit once we merge the other docs tickets for `mender-shell`.
